src/UnionSums.hs

```diff
@@ -32,20 +34,46 @@ unionSumTypes newNameStr typeNames =
     mkNewConstructors typeName = do
       info <- reify typeName
       case info of
-        (TyConI (DataD _ _ _ _ constructors _)) ->
-          either fail return $ mapM (modConstructor typeName) constructors
+        (TyConI dec) ->
+          either fail return $ mapM (modConstructor typeName) $ getConstructors dec
         _ -> fail $ nameBase typeName ++ " must be a sum type"
 
     modConstructor :: Name -> Con -> Either String Con
     modConstructor typeName (NormalC conName args) =
-        (\n -> (NormalC (mkName $ n ++ newNameStr) args)) <$>
+        (\n -> NormalC (mkName n) args) <$>
         note "Constructor name missing type name suffix"
-          (stripNameSuffix typeName conName)
+          (changeSuffix (nameBase typeName) newNameStr (nameBase conName))
     modConstructor _ _ = fail "Unrecognised constructor pattern"
 
-    stripNameSuffix :: Name -> Name -> Maybe String
-    stripNameSuffix suf = stripSuffix (nameBase suf) . nameBase
+toMainTypeConversion :: Name -> Name -> Q [Dec]
```

Comment:
"main type" is a bit of a strange concept here. When I read "main" I think of "main function", rather than "my most prominent thing". Perhaps something like `mkConverter` might be the way to go?